In [48]:
import random

In [49]:
def gcd(a, b):
    """
    Алгоритм Евклида.
    
    Параметры:
    ----------
    a: int
        первое число
    b: int
        второе число
    """
    if b == 0:
        return a
    return gcd(b, a % b)

In [50]:
def egcd(a, b):
    """
    Расширенный алгоритм Евклида.
    
    Параметры:
    ----------
    a: int
        первое число
    b: int
        второе число
    """
    if b == 0:
        return (1, 0, a)
    (x, y, d) = egcd(b, a % b)
    return (y, x - (a // b) * y, d)

In [51]:
def primality(N):
    """
    Тест Ферма (проверка числа на простоту).
    
    Параметры:
    ----------
    N: int
    """
    flag = 0
    k = 3
    if N == 1 or N == 4:
        return False
    elif N == 2 or N == 3:
        return True
    for _ in range(k):
        a = random.randint(1, N-1)
        if (a**(N-1) % N != 1):
            flag = 1
            break
    if flag == 0:
        #print('This is a prime number')
        return True
    else:
        #print('This is a composite number')
        return False

Получим список всех простых чисел в диапазоне от 1 до 1000:

In [87]:
ls_prime = []
k = 0
for i in range(1, 1001):
    if(primality(i)):
        k += 1
        ls_prime.append(i)
print(k)

170


In [88]:
print(ls_prime)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 561, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 703, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]


In [89]:
ls_prime2 = []
k = 0
for i in range(1, 1001):
    if(primality(i)):
        k += 1
        ls_prime2.append(i)
print(k)

168


In [90]:
print(ls_prime2)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]


In [92]:
for i in ls_prime:
    if not(i in ls_prime2):
        print(i)

561
703


In [109]:
primality(703)

True

И если число 561 - число Кармайкла(при любом случайном числе проходит проверку Ферма), то при одном из вызовов функции 703 также ошибочно функция выдала за простое. Это происходит, так как условие в Малой теореме Ферма не является необходимым(хотя является достаточным) для проверки числа на простоту. 

Результат работы функции egcd на примере чисел 346 и 42:

In [54]:
a, b = 346, 42
x, y, d = egcd(a, b)

In [55]:
print(x*a + y*b)

2


In [56]:
print(d)

2
